published_date/published_date.py

```diff
@@ -0,0 +1,109 @@
+import logging, os
+import time
+import json
+from json.decoder import JSONDecodeError
+from requests_futures.sessions import FuturesSession
+from concurrent.futures import as_completed
+
+
+logger = logging.getLogger(__name__)
+
+
+class FetchPublishedDate:
+
+    def __init__(self, canvas_url, canvas_token, workers, canvas_ids):
+        self.canvas_url = canvas_url,
+        self.canvas_token = canvas_token,
+        self.canvas_ids = canvas_ids
+        self.workers = workers
+        self.published_course_date = {}
+        self.published_course_next_page_list = []
+
+    def get_next_page_url(self, response):
+        """
+        get the next page url from the Http response headers
+        :param response:
+        :type response: requests
+        :return: next_page_url
+        :rtype: str
+        """
+        logging.debug(self.get_next_page_url.__name__ + '() called')
+        results = response.result().links
+        if not results:
+            logging.debug('The api call do not have Link headers')
+            return None
+
+        if 'next' in results:
+            url_ = results['next']['url']
+            self.published_course_next_page_list.append(url_)
+            logger.info(f"Pagination size {len(self.published_course_next_page_list)} for published_at date")
+
+    def published_date_resp_parsing(self, response):
+
+        logger.info("published_date_resp_parsing Call")
+
+        if response is None:
+            logger.info(f"Published course date response is None ")
+            return
+
+        start_time = time.time()
+
+        logger.info(f"published courses date collected so far : {len(self.published_course_date)}")
+        status = response.result().status_code
+        published_date_found = False
+        if status != 200:
+            logger.info(f"Response not successful with status code {status} due to {response.result().text}")
+            return
+
+        try:
+            audit_events = json.loads(response.result().text)
+        except JSONDecodeError as e:
+            logger.info(f"Error in parsing the response {e.message}")
+            return
+
+        if not audit_events:
+            logger.info(f"Response for fetching published date is empty {audit_events}")
+            return
+
+        events = audit_events['events']
+
+        for event in events:
+            if event['event_type'] == 'published':
+                course_id = event['links']['course']
+                published_date_found = True
+                self.published_course_date.update({course_id: event['created_at']})
+                logger.info(f"Published Date {event['created_at']} for course {course_id}")
+                break
+        if not published_date_found:
+            self.get_next_page_url(response)
```

Comment:
So if there were multiple `published` events (i.e. an instructor published and un-published the course multiple times), this logic would just find the first one, right? Did you look to see if there were ever multiple? Also, it's conceivable that one published event could be on one page and another published event could be on another...

Reply:
I asked this question before in slack and it was decided that to get the latest published date. And The way the canvas course audit logs work they send the audit events in desc order

Reply:
Okay, I do remember that conversation. Can you please put a comment here that the audit logs are in descending order? Otherwise, this logic isn't clear to me.

Reply:
Yes and I tested if ever the course has more event logs that we can paginate and get more data

Reply:
Thanks!